hail/python/test/hail/expr/test_expr.py

```diff
@@ -101,7 +101,23 @@ def test_operators(self):
             x33=(kt.a == 0) & (kt.b == 5),
             x34=(kt.a == 0) | (kt.b == 5),
             x35=False,
-            x36=True
+            x36=True,
+            x37=kt.e > "helln",
+            x38=kt.e < "hellp",
+            x39=kt.e <= "hello",
+            x40=kt.e >= "hello",
+            x41="helln" > kt.e,
+            x42="hellp" < kt.e,
+            x43="hello" >= kt.e,
+            x44="hello" <= kt.e,
+            x45=kt.f > [1, 2],
+            x46=kt.f < [1, 3],
+            x47=kt.f >= [1, 2, 3],
+            x48=kt.f <= [1, 2, 3],
+            x49=kt.f < [1.0, 2.0],
+            x50=kt.f > [1.0, 3.0],
+            x51=[1.0, 2.0, 3.0] <= kt.f,
+            x52=[1.0, 2.0, 3.0] >= kt.f,
```

Comment:
can we add another example:

```
x53=hl.tuple([True, 1.0]) < (1, 0)
```

(should be False)

Reply:
Well, great case. Because it triggered some tricky thing. Exception thrown.

```
    def _compare_op(self, op, other):
        other = to_expr(other)
        left, right, success = unify_exprs(self, other)
        if not success:
>           raise TypeError(f"Invalid '{op}' comparison, cannot compare expressions "
                            f"of type '{self.dtype}' and '{other.dtype}'")
E           TypeError: Invalid '<' comparison, cannot compare expressions of type 'tuple(bool, float64)' and 'tuple(int32, int32)'
```

Since `bool` < `int32` < `float64`, I expected the coerced type should be `tuple(int32, float64)`. The `TupleCoercer` seems not work as expected.

Reply:
Ah! Boolean -> int32 conversion is something we support explicitly for arithmetic, but not everywhere. I'd like to think through more cases before we open that up.

Let's make the test case 

```
x53=hl.tuple([1, 1.0]) < (1.0, 0)
```

Reply:
I think I can hotfix the `can_coerce` check with some ugly code:

```
    def can_coerce(self, t: HailType):
        import pdb; pdb.set_trace()
        if self.elements is None:
            return isinstance(t, ttuple)
        else:
            return (isinstance(t, ttuple)
                    and len(t.types) == len(self.elements)
                    and all(c.can_coerce(t_) or coercer_from_dtype(t_).can_coerce(hl.dtype(c.str_t)) for c, t_ in zip(self.elements, t.types)))
```

but i have not found a way to fix the `_coerce` method.

Reply:
ah wait, hmm...

Reply:
the problem is in unify_exprs:

```
    for t in types:
        c = expressions.coercer_from_dtype(t)
        if all(c.can_coerce(e.dtype) for e in exprs):
            return tuple([c.coerce(e) for e in exprs]) + (True,)
```

This is looking for one type that can coerce the rest. The correct thing is to recursively walk the structure of the types (tuple, struct, etc) and determine the coercer for each element. This is trickier and shouldn't be a part of this PR.

Instead, let's just change the check to:

```
x53=hl.tuple([1, 1.0]) < (1.0, 0.0)
```

and save the unify_exprs fix for another PR.

Reply:
sure. np.